evm-tests/jsontests/src/state.rs

```diff
@@ -426,22 +426,185 @@ pub fn test(
+	assert!(
+		check_exception,
+		"expected EmptyCaller exception for test: {name}"
+	);
+}
+
+/// Check call expected exception
+fn assert_call_exit_exception(spec: &ForkSpec, expect_exception: &Option<String>) {
+	if *spec == ForkSpec::Berlin {
+		if let Some(exception) = expect_exception.as_deref() {
+			let check_result = exception == "TR_TypeNotSupported";
+			assert!(check_result, "expected call exception");
+		}
+	} else {
+		assert!(expect_exception.is_none(), "unexpected call exception");
+	}
+}
+
 /// Check Exit Reason of EVM execution
-fn check_create_exit_reason(reason: &ExitReason, expect_exception: &Option<String>) -> bool {
-	if let Some(exception) = expect_exception.as_deref() {
-		if matches!(reason, ExitReason::Error(ExitError::CreateContractLimit)) {
-			let check_result = exception == "TR_InitCodeLimitExceeded"
-				|| exception == "TransactionException.INITCODE_SIZE_EXCEEDED";
+fn check_create_exit_reason(
+	reason: &ExitReason,
+	expect_exception: &Option<String>,
+	name: &str,
+) -> bool {
+	match reason {
+		ExitReason::Error(err) => {
+			if let Some(exception) = expect_exception.as_deref() {
+				match err {
+					ExitError::CreateContractLimit => {
+						let check_result = exception == "TR_InitCodeLimitExceeded"
+							|| exception == "TransactionException.INITCODE_SIZE_EXCEEDED";
+						assert!(
+							check_result,
+							"unexpected exception {exception:?} for CreateContractLimit error for test: {name}"
+						);
+						return true;
+					}
+					ExitError::MaxNonce => {
+						let check_result = exception == "TR_NonceHasMaxValue";
+						assert!(check_result,
+								"unexpected exception {exception:?} for MaxNonce error for test: {name}"
+						);
+						return true;
+					}
+					_ => {
+						panic!("unexpected error: {err:?} for exception: {exception}")
+					}
+				}
+			} else {
+				return false;
+			}
+		}
+		ExitReason::Fatal(err) => {
+			panic!("Unexpected error: {err:?}")
+		}
+		_ => {
 			assert!(
-				check_result,
-				"message: {exception}\nbut expected init code limit exceeded"
+				expect_exception.is_none(),
+				"Unexpected json-test error: {expect_exception:?}"
 			);
-			return true;
 		}
 	}
 	false
 }
 
+/// Check Exit Reason of EVM execution
+fn check_validate_exit_reason(
+	reason: &InvalidTxReason,
+	expect_exception: &Option<String>,
+	name: &str,
+) -> bool {
+	expect_exception.as_deref().map_or_else(
+		|| {
+			panic!("unexpected validation error reason: {reason:?}");
+		},
+		|exception| {
+			match reason {
+				InvalidTxReason::OutOfFund => {
+					let check_result = exception
+						== "TransactionException.INSUFFICIENT_ACCOUNT_FUNDS"
+						|| exception == "TR_TypeNotSupported"
+						|| exception == "TR_NoFunds"
+						|| exception == "TR_NoFundsX"
+						|| exception == "TransactionException.INSUFFICIENT_MAX_FEE_PER_BLOB_GAS";
+					assert!(
+						check_result,
+						"unexpected exception {exception:?} for OutOfFund  for test: {name}"
```

Comment:
Redundant space and below as well.
```suggestion
						"unexpected exception {exception:?} for OutOfFund for test: {name}"
```